tests/python_package_test/test_basic.py

```diff
@@ -243,6 +243,13 @@ def test_chunked_dataset_linear():
     valid_data.construct()
 
 
+def test_save_dataset_subset_and_load_from_file(tmp_path):
+    data = np.random.rand(100, 2)
+    ds = lgb.Dataset(data)
```

Comment:
Could you please try overriding the defaults (e.g. setting `max_bin: 7` or something) and check that those non-default values survive the round trip to disk?

I think that would increase our confidence that this is working as expected. Otherwise, I think a bug of the form "all parameter values are lost when writing to binary Dataset file" could make it through this test.

Reply:
I think that will raise an error like the one in #4902, but I'll try it and confirm here.

Reply:
Ok I see. I expected a PR called "include parameters" to test that parameters on either side of an operation had the same values.

I think maybe I got confused by the presence of writing and reading a binary file, and thought the issue was specific to storing a `Dataset` to disk. Is it like *"when taking a subset, not all parameters are copied from the reference Dataset to the subset...and this can show up as an error loading the Dataset from file"*?

If it is, then it would be great to be able to reach into the Dataset (on the C++ side, not the Python object) and check that attributes like `max_bin_` are the same and a non-default value. But I'm not sure how to do that without introducing a new `c_api` entrypoint.

So if adding the test I suggested does hit the error from #4904, then I think this test in its current state is ok. It is still an improvement that fixes a bug.

Reply:
I built the dataset with non default parameters and check that loading it with the same ones succeeds in 6a2fd1f.

Reply:
Ok nice! Thank you for that and for the explanation.

I tried running this PR's test code on latest `master`, and can see based on the error message that those non-default parameter values are being respected.

```python
import lightgbm as lgb
import numpy as np

data = np.random.rand(100, 2)
params = {'max_bin': 50, 'min_data_in_bin': 10}
ds = lgb.Dataset(data, params=params).construct()
ds.subset([1, 2, 3, 5, 8]).save_binary('subset.bin')
lgb.Dataset('subset.bin', params=params).construct()
```

> [LightGBM] [Fatal] Dataset was constructed with parameter max_bin=32649. It cannot be changed to **50** when loading from binary file.